Avoid using currentConstraint as predicate for getLayouts call

haozhun opened the pull request:
Fixes #11134

For the query mentioned in #11134: a query with 16 scans of a table with 10000 partitions (without any partition filter). The query's planning time reduced by 4X, from 115 seconds to 25 seconds. The query's initial task update request size reduced by 600X, from 4.4MB (half of what was claimed in #11134 due to #11133) to 7KB.

kokosing commented:
Would it be possible to merge this after: https://github.com/prestodb/presto/pull/11164?

haozhun commented on presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java:
```diff
@@ -56,4 +60,28 @@ public boolean hasAllOutputs(TableScanNode node)
         return columns.containsAll(nodeColumnHandles);
     }
+
+    public static TupleDomain<ColumnHandle> computeEnforced(TupleDomain<ColumnHandle> tupleDomain, TupleDomain<ColumnHandle> unenforcedTupleDomain)
```
I'll add tests.

Yes, your interpretation is correct. Is there any confusion here on what the `tupleDomain` argument means? I tried to come up with more explicit names, but then figured that the semantics is sufficiently as is in the context of this method. Maybe I'm wrong. Do you have any recommendations?

kokosing replied:
I think `predicate` instead of `tupleDomain` is a better name, because it is a predicate we are trying to push down.

haozhun replied:
Done.

haozhun commented on presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java:
```diff
@@ -56,4 +60,28 @@ public boolean hasAllOutputs(TableScanNode node)
+    public static TupleDomain<ColumnHandle> computeEnforced(TupleDomain<ColumnHandle> tupleDomain, TupleDomain<ColumnHandle> unenforcedTupleDomain)
+    {
+        if (tupleDomain.isNone()) {
+            checkArgument(unenforcedTupleDomain.isNone() || unenforcedTupleDomain.isAll());
```
If the engine requests a connector provides a layout with a domain of "none". The connector can have two possible reactions, either:

1. It can provide an empty table layout, in which case there would be no unenforced domain
2. It cannot. In which case, it will tell the engine to help it filter everything out by providing a "false" unenforced domain.

I had this here as a sanity check to assert one of these two scenarios. It would be invalid argument otherwise. I'll put my answer above as comment in the code.

kokosing replied:
> If the engine requests a connector provides a layout with a domain of "none"

It should not happen. `com.facebook.presto.metadata.MetadataManager#getLayouts` will return an empty list for domain `none`.

haozhun commented on presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java:
```diff
@@ -56,4 +60,28 @@ public boolean hasAllOutputs(TableScanNode node)
+        ImmutableMap.Builder<ColumnHandle, Domain> enforcedDomainsBuilder = ImmutableMap.builder();
+        for (Map.Entry<ColumnHandle, Domain> entry : domains.entrySet()) {
+            ColumnHandle columnHandle = entry.getKey();
+            if (!unenforcedDomains.containsKey(columnHandle)) {
```
I don't think so. See `TupleDomain.normalizeAndCopy`.

haozhun commented on presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java:
```diff
@@ -56,4 +60,28 @@ public boolean hasAllOutputs(TableScanNode node)
+        Map<ColumnHandle, Domain> domains = tupleDomain.getDomains().get();
+        Map<ColumnHandle, Domain> unenforcedDomains = unenforcedTupleDomain.getDomains().get();
+        ImmutableMap.Builder<ColumnHandle, Domain> enforcedDomainsBuilder = ImmutableMap.builder();
+        for (Map.Entry<ColumnHandle, Domain> entry : domains.entrySet()) {
```
`predicateDomain` would be `entry.getValue()`, not `entry`?

kokosing replied:
yes, but `predicateDomain.getValue()` seems logical. Even`predicateDomain.getKey()` looks good in this context IMO.

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/BeginTableWrite.java:
```diff
@@ -173,23 +175,24 @@ private PlanNode rewriteDeleteTableScan(PlanNode node, TableHandle handle)
         {
             if (node instanceof TableScanNode) {
                 TableScanNode scan = (TableScanNode) node;
+                TupleDomain<ColumnHandle> originalEnforcedConstraint = scan.getEnforcedConstraint();
```
I don't understand why you would want to inline this given this variable has a meaningful name and multiple usages.

kokosing replied:
Because the name is no shorter than the code it wraps. Code to me is better because it is more flexible. Anyway, up to you.

haozhun commented on presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java:
```diff
@@ -56,4 +60,28 @@ public boolean hasAllOutputs(TableScanNode node)
+        Map<ColumnHandle, Domain> unenforcedDomains = unenforcedTupleDomain.getDomains().get();
+        ImmutableMap.Builder<ColumnHandle, Domain> enforcedDomainsBuilder = ImmutableMap.builder();
+        for (Map.Entry<ColumnHandle, Domain> entry : domains.entrySet()) {
+            ColumnHandle columnHandle = entry.getKey();
```
In general, I prefer to assign `entry.getKey` and `entry.getValue` to variables so that they can have a name. But I also acknowledge that in Presto, we prefer strongly toward inlining, as a result, I only do this if a variable is used multiple times, as is the case here.

haozhun commented on presto-main/src/main/java/com/facebook/presto/server/remotetask/HttpRemoteTask.java:
```diff
@@ -89,9 +89,9 @@
 import static com.google.common.util.concurrent.Futures.immediateFuture;
```
I included it here because the end goal of this PR is to reduce this number. I had some measurement, but I forgot to put the result in the description of this PR.

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java:
```diff
@@ -183,56 +178,30 @@ public Result apply(TableScanNode tableScanNode, Captures captures, Context cont
     {
-        Expression deterministicPredicate = filterDeterministicConjuncts(predicate);
-        DomainTranslator.ExtractionResult decomposedPredicate = DomainTranslator.fromPredicate(
+        return AddExchanges.pickTableLayout(
```
I'll swap it the other way around for now. I can see your concern, but I don't think a separate class would be much better. After all, it's closely related to these two classes, and not something generic.

I chose to put it in `AddExchange` initially because that was the more up-to-date version. However, `AddExchange` is such a behemoth that having `PickTableLayout` in there can indeed be concerning.

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java:
```diff
@@ -1345,6 +1268,99 @@ else if (match2.isPresent()) {
         };
     }
 
+    public static List<PlanNode> pickTableLayout(
```
I'll rename

haozhun commented:
@martint ready for review

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java:
```diff
@@ -183,55 +205,157 @@ public Result apply(TableScanNode tableScanNode, Captures captures, Context cont
+        Map<ColumnHandle, Symbol> assignments = ImmutableBiMap.copyOf(node.getAssignments()).inverse();
+
+        Constraint<ColumnHandle> constraint;
+        if (pruneWithPredicateExpression) {
```
We do this because invoking the expression interpreter is currently very expensive (it used to be fast, fwiw). Add a comment explaining that and a TODO to get rid of this branch once we optimize the interpreter to make it cheap enough to call always.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java:
```diff
@@ -167,6 +167,11 @@ private boolean arePlansSame(FilterNode filter, TableScanNode tableScan, PlanNod
             }
 
             TableScanNode rewrittenTableScan = (TableScanNode) rewrittenFilter.getSource();
+
```
Typo in commit message: "guranteed"

martint commented on presto-main/src/main/java/com/facebook/presto/server/remotetask/HttpRemoteTask.java:
```diff
@@ -495,22 +495,23 @@ private synchronized void sendUpdate()
-        if (sendPlan.get()) {
-            fragment = Optional.of(planFragment);
-        }
+        Optional<PlanFragment> fragment = sendPlan.get() ? Optional.of(planFragment) : Optional.empty();
```
Unrelated change

martint commented on presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java:
```diff
@@ -14,13 +14,17 @@
 package com.facebook.presto.metadata;
```
> More details can be found in issue 11134.

Use # when referencing the issue so github turns it into a link.

> [...] this commit reduces flexibility of unenforcedConstraint
that a connector can return in getTableLayouts by requiring connector.

Sentence is incomplete.

martint commented on presto-main/src/main/java/com/facebook/presto/metadata/TableLayoutResult.java:
```diff
@@ -56,4 +60,54 @@ public boolean hasAllOutputs(TableScanNode node)
+            //   * There would be no unenforced predicate, i.e., unenforced predicate is TupleDomain.all().
+            //   * The predicate was successfully enforced. Enforced predicate would be same as predicate: TupleDomain.none().
+            // 2. The connector can't/won't.
+            //   * The connector would tell the engine to put a filter on top of the scan, i.e., unenforced predicate is TupleDomain.non().
```
Typo: non()

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TableScanNode.java:
```diff
@@ -40,23 +39,26 @@
     private final List<Symbol> outputSymbols;
     private final Map<Symbol, ColumnHandle> assignments; // symbol -> column
 
+    private final Optional<TableLayoutHandle> tableLayout;
```
Unrelated change?

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TableScanNode.java:
```diff
@@ -40,23 +39,26 @@
+
     @JsonCreator
-    public TableScanNode(
+    public static TableScanNode jsonCreator(
```
What's a "jsonCreator"? Call this "newInstance", instead.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TableScanNode.java:
```diff
@@ -74,23 +76,43 @@ public TableScanNode(
+            Optional<TableLayoutHandle> tableLayout,
+            TupleDomain<ColumnHandle> currentConstraint,
+            TupleDomain<ColumnHandle> enforcedConstraint,
+            boolean internal)
```
This flag here is a code smell. Looking at the rest of the commit, it seems to be used only in the static factory method above. Move the check there, instead.

haozhun commented:
Addressed comments

----------------------

When I was reading this code, `TableScanNode` was a pretty confusing class. The constructor of `TableScanNode` had to omit a bunch of sanity checks to make the json creator work, making the invariants not obvious.

I opted to fix this by having a "internal" boolean parameter on the private constructor. A "dummy" Void parameter could achieve the same outcome.

An alternative is to have the json creator not delegate to another constructor, this will solve the above issue. But it carries two problems with it:
1. The json deserializer here should be private. That's not possible due to Jackson constraint. However, Jackson allows using factory method. With this alternative, factory methods doesn't work because factories have to delegate to constructors.
2. Minor code duplication.

I don't feel very strongly here. As a result, I took the above alternative that you suggested.
